src/v/arch/manifest.cc

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright 2021 Vectorized, Inc.
+ *
+ * Licensed as a Redpanda Enterprise file under the Redpanda Community
+ * License (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * https://github.com/vectorizedio/redpanda/blob/master/licenses/rcl.md
+ */
+
+#include "arch/manifest.h"
+
+#include "arch/error.h"
+#include "arch/logger.h"
+#include "bytes/iobuf.h"
+#include "bytes/iobuf_istreambuf.h"
+#include "bytes/iobuf_ostreambuf.h"
+#include "hashing/murmur.h"
+#include "model/fundamental.h"
+#include "model/metadata.h"
+#include "storage/ntp_config.h"
+
+#include <seastar/core/coroutine.hh>
+
+#include <boost/algorithm/string.hpp>
+#include <boost/property_tree/json_parser.hpp>
+#include <boost/property_tree/ptree.hpp>
+
+#include <algorithm>
+#include <array>
+
+namespace arch {
+
+manifest::manifest()
+  : _ntp()
+  , _rev() {}
+
+manifest::manifest(model::ntp ntp, model::revision_id rev)
+  : _ntp(std::move(ntp))
+  , _rev(rev) {}
+
+remote_manifest_path manifest::get_manifest_path() const {
+    static constexpr std::array<uint32_t, 15> bins = {{
```

Comment:
do we want these bins to be prime numbers instead? from some march archive.

Reply:
Not if we assume that the hash distribution is even.  In this case we just need to split the prefixes into N more or less equally sized bins. The boundaries should really be 0x10000000, 0x20000000, etc.  I choose this ones because I thought that they'll be easier to spot while examining the content of the bucket. If you see something like 0x77777777 this means that it contains manifests (it can also contains segments btw). I should just change this to use equally sized bins. Otherwise it 's just confusing.

Reply:
ah! makes sense.

Reply:
I think we  should copy-paste this comment above that array.